### DISCUSSION TITLE
[Usage question] Does libvips support getting phash of an image?

### DISCUSSION THREAD

**Discussion by ngocdaothanh** [libvips/libvips#2723] [Q&A]
Does libvips support getting phash of an image, and diffing two images?

**Comment by jcupitt** [comment]
Hi @ngocdaothanh,

Not directly, but it's easy to implement perceptual hashing functions using libvips. For example, in ruby you could write (untested):

```ruby
# convert to mono
image = image.colourspace("b-w")

# shrink to an 8x8 image with a simple box filter
image = image.shrink(image.width / 8.0, image.height / 8.0)

# set pixels to 255 or 0 if they are greater or less than the image average
image = image > image.avg

values = image.to_a
```

`values` will be something like `[[[0], [255] ...`. Take the top bit of each of the 64 numbers and form a 64-bit uint, and that's your hash.

You can find the distance between two hashes by EORing them and counting the set bits.

@Nakilon has a much fancier perceptual hash here:

https://github.com/Nakilon/dhash-vips

**Comment by richardmonette** [comment]
With apologies for bumping the old thread, in case it's helpful to anyone, I noticed https://github.com/khasinski/phash-rb which uses libvips to generate a phash. May be a useful example of how to make a phash that could be ported to other languages as needed.
